### PR TITLE
Alter Admin initialization to admit plugin-defined handlers

### LINKED ISSUE #98: remove vestigal `/admin` UI
Functionality is now in /, and things have changed sufficiently that this endpoint doesn't even do anything, just shows a empty UI.

### PR DESCRIPTION
**Pull request by olix0r**
#### Problem

There's no mechanism for plugins to install endpoints onto the admin service.
There should be.

#### Solution

Introduce the `Admin.WithHandlers` trait, which allows admin initialization to
detect modules that expose handlers.

Refactor Admin initialization to compose handlers during initialization.
Previously, admin classes were subtyped to extend functionality. Now, admin
endpoints are concatenated during initialization so that an arbitrary number of
endpoints may be installed without comabting the type system.

Furthermore, the `Namerd` and `Linker` types now have
configured-but-not-yet-initialized `Admin` values (rather than `AdminConfig`),
so that Admin initialization is congruent with other subsystems.

Finally, remove unused/unwanted parts of admin interface to fix #98.

### CODE REVIEW

File: admin/src/main/scala/io/buoyant/admin/Admin.scala [hunk 1]
```diff
@@ -1,65 +1,112 @@
...
     "/admin/per_host_metrics.json" -> HttpMuxer
   )
 
-  def allRoutes = twitterServerRoutes ++ metricsRoutes
+  /** Generate an index of the provided handlers */
+  private def indexHandlers(handlers: Handlers): Handlers = {
+    val paths = handlers.map { case (p, _) => p }.sorted.distinct
+    val index = new IndexTxtHandler(paths)
+    Seq(
+      "/admin/index.txt" -> index,
```
Comment (adleong) [inline admin/src/main/scala/io/buoyant/admin/Admin.scala]: What are these text index routes for?

File: admin/src/main/scala/io/buoyant/admin/Admin.scala [hunk 2]
```diff
@@ -1,65 +1,112 @@
...
+  }
+
+  def getHandlers(obj: Any): Handlers =
+    obj match {
+      case wh: WithHandlers => wh.adminHandlers
+      case _ => Nil
+    }
+
+  def collectHandlers(objs: Seq[Any]): Handlers =
+    objs.flatMap(getHandlers(_))
```
Comment (olix0r) [inline admin/src/main/scala/io/buoyant/admin/Admin.scala]: not required, but it signals (to me) that there's a function allocation here.  Got this convention from finagle.

File: admin/src/main/scala/io/buoyant/admin/Admin.scala (hunk 1 above)
Comment (olix0r) [inline admin/src/main/scala/io/buoyant/admin/Admin.scala]: just so you have a way to list all of the admin endpoints without guessing or reading the code (a problem i personally have a lot)

File: linkerd/admin/src/main/scala/io/buoyant/linkerd/admin/LinkerdAdmin.scala [hunk 3]
```diff
@@ -13,38 +14,51 @@ import io.buoyant.linkerd.Linker
...
-    linker.routers.find(_.label == label).map { router =>
-      router.params[DstBindingFactory.Namer].interpreter
-    }.getOrElse(NameInterpreter)
-
-  override def allRoutes = super.allRoutes ++ linkerdAdminRoutes
+  def apply(lc: Linker.LinkerConfig, linker: Linker): Admin.Handlers =
+    static ++ config(lc) ++
+      boundNames(linker.namers.map { case (_, n) => n }) ++
+      delegator(linker.routers) ++
+      Admin.collectHandlers(linker.namers ++ linker.routers ++ linker.telemeters)
```
Comment (adleong) [inline linkerd/admin/src/main/scala/io/buoyant/linkerd/admin/LinkerdAdmin.scala]: what about handlers on non-top-level plugins (like identifiers, for example)?

Comment (olix0r) [inline linkerd/admin/src/main/scala/io/buoyant/linkerd/admin/LinkerdAdmin.scala]: TBD.  we'll have to plumb this through into the router, presumably

**Comment by adleong** [comment]
:star: 🉐 

[Approved with PullApprove](https://pullapprove.com/BuoyantIO/linkerd/pull-request/667/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=BuoyantIO/linkerd)

**Comment by olix0r** [comment]
no js changes, rolling forward with this

**Pull request merged**